README.md
Install darkice with mp3 support on a raspberry pi

## Requirements
A raspberry pi with chef installed

## Usage

###Attributes
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td colspan="4"><b>Default recipe</b></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['package_url']</tt></td>
    <td>String</td>
    <td>Remote url to download a precompiled package</td>
    <td><tt>empty</tt></td>
  </tr>
  <tr>
    <td colspan="4"><b>Source recipe</b></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['platform']</tt></td>
    <td>String</td>
    <td>Repository platform</td>
    <td><tt>armhf</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['uri']</tt></td>
    <td>String</td>
    <td>Repository uri</td>
    <td><tt>http://mirrordirector.raspbian.org/raspbian/</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['distribution']</tt></td>
    <td>String</td>
    <td>Repository distribution</td>
    <td><tt>wheezy</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['version']</tt></td>
    <td>String</td>
    <td>The version name for the self compiled package</td>
    <td><tt>1.0-999~mp3+1</tt></td>
  </tr>
  <tr>
    <td colspan="4"><b>Configuration</b></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['input']['device']</tt></td>
    <td>String</td>
    <td>Input device to capture sound from</td>
    <td><tt>hw:0,0</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['input']['sample-rate']</tt></td>
    <td>String</td>
    <td>MP3 Sample rate</td>
    <td><tt>44100</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['input']['bits-per-sample']</tt></td>
    <td>String</td>
    <td>MP3 bits per sample</td>
    <td><tt>16</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['input']['channels']</tt></td>
    <td>String</td>
    <td>MP3 Channels</td>
    <td><tt>2</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['server']['hostname']</tt></td>
    <td>String</td>
    <td>Icecast Server hostname</td>
    <td><tt>raspberrypi.local</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['server']['port']</tt></td>
    <td>String</td>
    <td>Icecast Server port</td>
    <td><tt>8000</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['server']['password']</tt></td>
    <td>String</td>
    <td>Icecast Server password</td>
    <td><tt>password</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['stream']['bittrate']</tt></td>
    <td>String</td>
    <td>Stream bitrate</td>
    <td><tt>320</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['stream']['bitrate-mode']</tt></td>
    <td>String</td>
    <td>Stream bitrate mode</td>
    <td><tt>vbr</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['stream']['quality']</tt></td>
    <td>String</td>
    <td>Stream quality</td>
    <td><tt>1.0</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['stream']['format']</tt></td>
    <td>String</td>
    <td>Stream format</td>
    <td><tt>mp3</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['stream']['mount']</tt></td>
    <td>String</td>
    <td>Stream mount point</td>
    <td><tt>raspi</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['stream']['name']</tt></td>
    <td>String</td>
    <td>Stream name</td>
    <td><tt>Raspi</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['stream']['description']</tt></td>
    <td>String</td>
    <td>Stream description</td>
    <td><tt>DarkIce on RasPi</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['stream']['url']</tt></td>
    <td>String</td>
    <td>Stream url</td>
    <td><tt>http://raspberrypi.local</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['stream']['genre']</tt></td>
    <td>String</td>
    <td>Stream genre</td>
    <td><tt>turntable</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['stream']['public']</tt></td>
    <td>String</td>
    <td>Public or private Stream</td>
    <td><tt>no</tt></td>
  </tr>
  <tr>
    <td><tt>['darkice-rpi']['stream']['dump-file']</tt></td>
    <td>String</td>
    <td>Stream dump file</td>
    <td><tt>recording.mp3</tt></td>
  </tr>
</table>

#### darkice-rpi::default
To install darkice from a precompiled source include `darkice-rpi` in your node's `run_list`
with a remote package url:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[darkice-rpi]"
  ],
  "darkice-rpi": {
    "package_url": "https://"
  }
}
```

#### darkice-rpi::source
To install darkice from a source just include `darkice-rpi::source` in your node's `run_list`:

```json
{
  "name":"my_node",
  "run_list": [
    "recipe[darkice-rpi::source]"
  ]
}
```

### Contributing

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

### License and Authors

Authors: Tim Petter
Credits: https://stmllr.net/blog/live-streaming-mp3-audio-with-darkice-and-icecast2-on-raspberry-pi/

```
The MIT License (MIT)

Copyright (c) 2015 Tim Petter

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```


CHANGELOG.md
## darkice-ri CHANGELOG

### 0.2.0
- Add recipe to download precompiled package
- Update README.md

### 0.1.0

- First relase
  It is possible to install and configure darkice on a rapsberry pi.
  Currently only in combination with icecast2.


recipes/source.rb
#
# Cookbook Name:: darkice-rpi
# Recipe:: source
#
# Copyright 2015, Tim Petter
#
#
include_recipe 'apt'

apt_repository 'darkice-rpi' do
  uri node['darkice-rpi']['uri']
  distribution node['darkice-rpi']['distribution']
  components ['main', 'contrib', 'non-free', 'rpi']
  arch node['darkice-rpi']['platform']
  deb_src true
end

[
  'devscripts',
  'autotools-dev',
  'fakeroot',
  'dpkg-dev',
  'debhelper',
  'autotools-dev',
  'dh-make',
  'quilt',
  'ccache',
  'libsamplerate0-dev',
  'libpulse-dev',
  'libaudio-dev',
  'lame',
  'libjack-jackd2-dev',
  'libasound2-dev',
  'libtwolame-dev',
  'libfaad-dev',
  'libflac-dev',
  'libmp4v2-dev',
  'libshout3-dev',
  'libmp3lame-dev'
].each do |p|
  package p do
    action :install
  end
end

execute 'Get darkice source' do
  cwd '/usr/local/src'
  command 'apt-get source darkice'
end

template '/usr/local/src/darkice-1.0/debian/rules' do
  source 'rules.erb'
end

execute 'Change version to prevent override' do
  cwd '/usr/local/src/darkice-1.0'
  command "debchange -v #{ node['darkice-rpi']['version'] } mp3 support"
end

execute 'Build package' do
  cwd '/usr/local/src/darkice-1.0'
  command 'dpkg-buildpackage -rfakeroot -uc -b'
  not_if { ::File.exist?("/usr/local/src/darkice_#{ node['darkice-rpi']['version'] }_armhf.deb") }
end

execute 'Install package' do
  cwd '/usr/local/src'
  command "dpkg -i darkice_#{ node['darkice-rpi']['version'] }_armhf.deb"
end

template '/etc/darkice.cfg' do
  source 'darkice.cfg.erb'
end


attributes/defaults.rb
# Source attributes
default['darkice-rpi']['platform'] = 'armhf'
default['darkice-rpi']['uri'] = 'http://mirrordirector.raspbian.org/raspbian/'
default['darkice-rpi']['distribution'] = 'wheezy'

# Version attributes
default['darkice-rpi']['version'] = '1.0-999~mp3+1'

# Configuration attributes
default['darkice-rpi']['input']['device'] = 'hw:0,0'
default['darkice-rpi']['input']['sample-rate'] = '44100'
default['darkice-rpi']['input']['bits-per-sample'] = '16'
default['darkice-rpi']['input']['channels'] = '2'

# Server attributes
default['darkice-rpi']['server']['hostname'] = 'raspberrypi.local'
default['darkice-rpi']['server']['port'] = '8000'
default['darkice-rpi']['server']['password'] = 'password'

# Icecast attributes
default['darkice-rpi']['stream']['bitrate'] = '320'
default['darkice-rpi']['stream']['bitrate-mode'] = 'vbr'
default['darkice-rpi']['stream']['quality'] = '1.0'
default['darkice-rpi']['stream']['format'] = 'mp3'
default['darkice-rpi']['stream']['mount'] = 'raspi'
default['darkice-rpi']['stream']['name'] = 'Raspi'
default['darkice-rpi']['stream']['description'] = 'DarkIce on RasPi'
default['darkice-rpi']['stream']['url'] = 'http://raspberrypi.local'
default['darkice-rpi']['stream']['genre'] = 'turntables'
default['darkice-rpi']['stream']['public'] = 'no'
default['darkice-rpi']['stream']['dump-file'] = 'recording.mp3'
